### PR TITLE
ENH: add refnx to staged-recipes

### PR DESCRIPTION
**Pull request by andyfaff**
Hi,
I'm the author of refnx, and I'm hoping to add my package to conda-forge. Hopefully this PR is relatively straightforward.

### CODE REVIEW

**Comment by andyfaff** [comment]
I marked the refnx package as being not zip_safe. I've tested a `conda-build` on a windows box I have access to, and the conda package built and tested successfully. Hopefully the appveyor CI doesn't cause any problems this time.

**Comment by jakirkham** [comment]
Please take a look at the [`example`]( https://github.com/conda-forge/staged-recipes/blob/9e809495446e259acc79d63390e9de8a753a9757/recipes/example/meta.yaml ) and update accordingly.

**Comment by andyfaff** [comment]
I did modify the example. It's not clear to me why the appveyor stage is failing, a `conda-build` works on my machine. @jakirkham, does one of the [build logs](https://ci.appveyor.com/project/conda-forge/staged-recipes/build/1.0.7355/job/wjslvw2r1sx0aogq) provide a clearer idea to you what the problem is?

**Comment by jakirkham** [comment]
Sorry there seems to be some confusion. Please update the proposed recipe to match the example. For instance, `build.sh` and `bld.bat` should be dropped in favor of `build/script` in `meta.yaml`.

**Comment by andyfaff** [comment]
Can drop `build.sh` and `build.bat`. AFAICT there are no other differences between my `meta.yaml` and the example.

**Comment by andyfaff** [comment]
I think there is something weird going on with the conda build process on appveyor with this recipe and I can't figure it out by myself. `run_test.py` completes successfully, but the build process goes no further. From an appveyor log below it seems that there is a temporary directory that should be removed, but for some reason it's not actually removable. I'm not sure what is actually in that work directory. If it's the installed package, then could it be some files with the wrong permissions? Alternatively, given that the package has a couple of c extensions could it be that those c extensions have been imported during the tests, thus preventing the directory from being deleted? I'm clutching at straws here.


    (root) C:\Miniconda\conda-bld\test-tmp_dir>call activate.bat C:\Miniconda\envs\_test 
    ===== testing package: refnx-0.0.5-np110py27_0 =====
    import: u'refnx'
    running run_test.py
    Running unit tests for refnx
    NumPy version 1.10.4
    NumPy relaxed strides checking option: False
    NumPy is installed in C:\Miniconda\envs\_test\lib\site-packages\numpy
    Python version 2.7.12 | packaged by conda-forge | (default, Sep  8 2016, 14:45:16) [MSC v.1500 32 bit (Intel)]
    nose version 1.3.7
    .................................................................C:\Miniconda\envs\_test\lib\site-packages\scipy\optimize\zeros.py:173: RuntimeWarning: Tolerance of 5.57221468966e-08 reached
      warnings.warn(msg, RuntimeWarning)
    ....................................................
    ----------------------------------------------------------------------
    Ran 117 tests in 117.313s
    OK
    TEST END: refnx-0.0.5-np110py27_0
    Building  refnx-0.0.5-np111py27_0
    Traceback (most recent call last):
      File "C:\Miniconda\Scripts\conda-build-all-script.py", line 11, in <module>
        load_entry_point('conda-build-all==1.0.0', 'console_scripts', 'conda-build-all')()
      File "C:\Miniconda\lib\site-packages\conda_build_all\cli.py", line 90, in main
        b.main()
      File "C:\Miniconda\lib\site-packages\conda_build_all\builder.py", line 263, in main
        built_dist_location = self.build(meta, build_config)
      File "C:\Miniconda\lib\site-packages\conda_build_all\builder.py", line 204, in build
        return bldpkg_path(build.build(meta.meta))
      File "C:\Miniconda\lib\site-packages\conda_build_all\build.py", line 33, in build
        shutil.rmtree(conda_build.source.WORK_DIR)
      File "C:\Miniconda\lib\shutil.py", line 247, in rmtree
        rmtree(fullname, ignore_errors, onerror)
      File "C:\Miniconda\lib\shutil.py", line 247, in rmtree
        rmtree(fullname, ignore_errors, onerror)
      File "C:\Miniconda\lib\shutil.py", line 247, in rmtree
        rmtree(fullname, ignore_errors, onerror)
      File "C:\Miniconda\lib\shutil.py", line 252, in rmtree
        onerror(os.remove, fullname, sys.exc_info())
      File "C:\Miniconda\lib\shutil.py", line 250, in rmtree
        os.remove(fullname)
    WindowsError: [Error 5] Access is denied: 'C:\\Miniconda\\conda-bld\\work\\.git\\objects\\pack\\pack-889e8c6e9e83694873e444c395a1b2318e86ff15.idx'
    Command exited with code 1

File: recipes/refnx/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+source:
+  git_url: https://github.com/refnx/refnx.git
+  git_rev: 404eceaff24d6b3ae116851dba25beb8069b0092
+
+build:
+  number: 0
+  script: python setup.py install --single-version-externally-managed --record record.txt
+
+requirements:
+  build:
```
Comment (jakirkham) [inline recipes/refnx/meta.yaml]: Please add `toolchain`.

File: recipes/refnx/run_test.py [hunk 2]
```diff
@@ -0,0 +1,9 @@
+import sys
+
+import refnx
+import refnx.analysis
+import refnx.util
+import refnx.dataset
+import refnx.reduce
```
Comment (jakirkham) [inline recipes/refnx/run_test.py]: Please add these to the import tests.

File: recipes/refnx/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+  home: https://github.com/refnx/refnx
+  license: BSD 3-Clause
+  license_file: LICENSE
+  summary: 'Neutron and X-ray Reflectometry Analysis'
+  doc_url: https://refnx.github.io
+
+extra:
+  recipe-maintainers:
+    - andyfaff
+
```
Comment (jakirkham) [inline recipes/refnx/meta.yaml]: nit: Please drop extra newline. One terminal newline is sufficient.

Comment (andyfaff) [inline recipes/refnx/meta.yaml]: ok

File: recipes/refnx/run_test.py (hunk 2 above)
Comment (andyfaff) [inline recipes/refnx/run_test.py]: I'm going to remove all these imports, only `import refnx` is needed for the test to run.
I do have a question here though, from conda build documentation one is supposed to list imports present in `run_test.py` in the yaml file. However, it's not explained why. I think having the answer to this question in FAQ would be helpful.

File: recipes/refnx/meta.yaml (hunk 1 above)
Comment (andyfaff) [inline recipes/refnx/meta.yaml]: Added. What does the toolchain package do? It's not clear why I have to add this, it's not present in the current recipe example. Could the answer to this be added to an FAQ?

**Comment by andyfaff** [comment]
Ok, so all CI tests pass now. Please let me know if there are any other changes I need to make.

**Comment by andyfaff** [comment]
Squashed commits and rebased on master.

File: recipes/refnx/meta.yaml [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+  run:
+    - python
+    - numpy x.x
+    - scipy
+    - xlrd
+    - pandas
+    - lmfit
+    - emcee
+    - uncertainties
+    - h5py
```
Comment (jakirkham) [inline recipes/refnx/meta.yaml]: Does this require Qt in some form? Seeing `.ui` files and PySide's imports.

**Comment by andyfaff** [comment]
There is some WIP in the 'motofit' directory that provides a GUI for data
analysis. However, this isn't being installed yet, it's way off being
ready. The setup.py script doesn't install this code, so there is no
requirement for Qt, or PySide.

File: recipes/refnx/meta.yaml [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+
+test:
+  requires:
+    - nose
+
+  imports:
+    - sys
```
Comment (jakirkham) [inline recipes/refnx/meta.yaml]: Wouldn't bother with `sys`. Just imports from your package should be fine.

File: recipes/refnx/meta.yaml (hunk 1 above)
Comment (jakirkham) [inline recipes/refnx/meta.yaml]: It configures the compilers used. Mainly this is of value on Mac particularly with C++. It works by sourcing some environment variables before the build. Please see the [feedstock]( https://github.com/conda-forge/toolchain-feedstock ) for more details.

Yeah, `simplejson` probably should have it as it is being compiled.

Yes, it probably should be in the FAQ.

Please feel free to open a PR for these.

Comment (jakirkham) [inline recipes/refnx/meta.yaml]: Sorry, forgot about that. Thanks for the reminder.

**Comment by jakirkham** [comment]
Thanks @andyfaff.

**Comment by jakirkham** [comment]
Soon you should get an email that invite you to join conda-forge, @andyfaff. Once accepted you will be added to a team with the same name as this recipe. Those will give you permissions on the feedstock (repo) for this recipe.

Make sure when proposing any change that you go through the typical GitHub workflow of forking the feedstock and making changes in your fork that you PR back. Once merged CIs will build and deploy any changes you make.

Please let us know if you have any questions and welcome to conda-forge. :)

**Pull request merged**